Detections/AzureDiagnostics/KeyvaultMassSecretRetrieval.yaml

```diff
@@ -1,58 +1,72 @@
-  | extend ResultType = columnifexists("ResultType", "None"), identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g = columnifexists("identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g", "None")
-  | where ResultType !~ "None" and isnotempty(ResultType)
-  | where identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g !~ "None" and isnotempty(identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g)
-  | where ResourceType =~ "VAULTS" and ResultType =~ "Success"
-  | where OperationName in (OperationList) 
-  | summarize count() by identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g, OperationName
-  | where count_ > EventCountThreshold  
-  | join (
-  AzureDiagnostics
-  | where not((identity_claim_appid_g in (Allowedappid)) and OperationName == 'VaultGet')
-  | extend ResultType = columnifexists("ResultType", "NoResultType")
-  | extend requestUri_s = columnifexists("requestUri_s", "None"), identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g = columnifexists("identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g", "None")
-  | extend id_s = columnifexists("id_s", "None"), CallerIPAddress = columnifexists("CallerIPAddress", "None"), clientInfo_s = columnifexists("clientInfo_s", "None")
-  | where ResultType !~ "None" and isnotempty(ResultType)
-  | where identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g !~ "None" and isnotempty(identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g)
-  | where id_s !~ "None" and isnotempty(id_s)
-  | where CallerIPAddress !~ "None" and isnotempty(CallerIPAddress)
-  | where clientInfo_s !~ "None" and isnotempty(clientInfo_s)
-  | where requestUri_s !~ "None" and isnotempty(requestUri_s)
-  | where OperationName in~ (OperationList)   
-  ) on identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g 
-  | summarize EventCount=sum(count_), StartTimeUtc=min(TimeGenerated), EndTimeUtc=max(TimeGenerated), TimeTriggered=makelist(TimeGenerated),OperationNameList=make_set(OperationName), RequestURLList=make_set(requestUri_s), CallerIPList = make_set(CallerIPAddress),  CallerIPMax= arg_max(CallerIPAddress,*) by ResourceType, ResultType, Resource, id_s, identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g, clientInfo_s
-  | extend timestamp = EndTimeUtc, IPCustomEntity = CallerIPMax, AccountCustomEntity = identity_claim_http_schemas_microsoft_com_identity_claims_objectidentifier_g
+  let AllowedAppId = dynamic(["509e4652-da8d-478d-a730-e9d4a1996ca4","8cae6e77-e04e-42ce-b5cb-50d82bce26b1"]);
+  let OperationList = dynamic(["SecretGet", "KeyGet", "VaultGet"]);
+  let Retrievals = materialize(
+      AzureDiagnostics
+      | where OperationName in (OperationList) and ResourceType =~ "VAULTS"
```

Comment:
can you please add TimeFilter since we are using materialize dont want to hold all the records in memory.. It was missing in original query too.

Reply:
@ashwin-patil this Scheduled rule has a query frequency and query period of 1d, if I am not mistaken, this query would only have access to data from the last day by the query scheduler, in that case the "TimeFilter" would be already applied, is this correct?

I am going to change the flow of the query a little bit, use the ```as``` operator, and remove ```materialize``` functions. Please, wait a bit.